### DISCUSSION TITLE
SDK - hooks on error

### DISCUSSION THREAD

**Discussion by pierre-H** [directus/directus#9660] [Feature Requests]
It would be great to have hook when the API return an error so we can execute a code on every error (logs, alerts etc.)

**Comment by rijkvanzanten** [comment]
Any thoughts on the ideal structure for such a thing? 

We could do it through an event emitter kinda setup:

```js
const directus = new Directus();

directus.on('error', (err) => {
  // do something
});
```

Should this then still throw where you're using the method?

```js
const directus = new Directus();

directus.on('error', (err) => {
  // do something
});

await directus.items('articles').readOne('broken ID');
// Should this crash or fail silently as it's handled in ".on('error')"?
```

**Reply by pierre-H** [reply]
The event emitter is a great idea !

For the broken id, fif the api returns a 404, it should be handled by the hook also.

**Reply by rijkvanzanten** [reply]
> it should be handled by the hook also.

I agree 👍🏻 With that in mind, should we design for the case where you might want to handle one error manually? 🤔 In other words: should you be able to have a global error handler for when you don't have an error handler?

**Reply by pierre-H** [reply]
I believe we should keep the `catch` for every request so we can handle error for specific calls.

Maybe we should have :
```tsx
directus.on('error', (err) => {
    console.log('before the catch');
});

directus.on('error:after', (err) => {
    console.log('after the catch');
});
```

**Reply by rijkvanzanten** [reply]
Hmm yeah, but in that case, the following will crash the process:

```js
const directus = new Directus();

directus.on('error', (err) => {
  // do something
});

await directus.items('articles').readOne('broken ID');
// process crashes with uncaught exception from promise
```

🤔

**Reply by pierre-H** [reply]
It is clear that using `await` forces the use of a `catch`.
We can set aside the `error:after` hook for the simplicity of the structure.

**Comment by rijkvanzanten** [comment]
Heya! Thank you for taking the time to submit this request! 

It has been over **90 days**, and this discussion has not received at least **15 votes** from the community. This means that we don't feel like there's enough community interest to warrant further R&D into this topic at this time. :ice_cube:

This request will now be closed to keep our discussions tidy. Please reach out if you have any questions!

_For more information, see our [Feature Request Process](https://docs.directus.io/contributing/feature-request-process.html)._
